### PR TITLE
Add predicate capturing results of test runs

### PR DESCRIPTION
**Pull request by adityasaky**

### CODE REVIEW

**Comment by TomHennen** [comment]
FYI one of my colleagues who has done work in this area is going to take a look and provide some comments. Stay tuned.

File: spec/predicates/test-results.md
Comment (clmccarthy5) [inline spec/predicates/test-results.md:15]: Out of curiosity, do you have immediate use-cases in mind?

File: spec/predicates/test-results.md
Comment (clmccarthy5) [inline spec/predicates/test-results.md:37]: What's the rationale for making the configuration reference optional?

It seems that having only the result with no indication of which tests were run may not be very helpful (e.g. it'd be good to know if the configured tests were from trusted config or untrusted config). I suppose this relates to my question above, about the potential use-cases for the predicate

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:15]: I'm working with an adopter to deploy in-toto and I started sketching this out based on their test framework.

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:37]: Continuing from comment about implementer, I made the configuration optional right now but I'm waiting to hear what parts they're willing to share with the results. I think there's value with configurations as well, so I'm okay to change this.

Also, do you have any thoughts on other things we'd include here?

Comment (clmccarthy5) [reply spec/predicates/test-results.md:37]: SGTM. Re: other things to include, I can't think of anything immediately -- IMO the reference to the test configuration and link to the test execution are the most important bits. It might be useful to consider the use-cases more explicitly, i.e. who is the consumer of test attestations, to see if any relevant info is missing (but feel free to ignore if there aren't concrete use-cases yet).

Comment (adityasaky) [reply spec/predicates/test-results.md:37]: Done!

Comment (adityasaky) [reply spec/predicates/test-results.md:59]: I just added some verification scenarios. Let me know if these make sense!

Comment (clmccarthy5) [reply spec/predicates/test-results.md:59]: Those all LGTM, thanks for the additional details!

**Review by clmccarthy5 — APPROVED** [review]

**Review by TomHennen — APPROVED** [review]
Thanks Aditya!

**Comment by adityasaky** [comment]
Thanks for the reviews! I've squashed my commits now.

File: spec/predicates/test-results.md
Comment (joshuagl) [inline spec/predicates/test-results.md:7]: we should update this before merging :-)

File: spec/predicates/test-results.md
Comment (joshuagl) [inline spec/predicates/test-results.md:31]: This is the first predicate to reference an inspection, should we be avoiding in-toto specification terms in the attestations repository (which aims to be policy engine agnostic)?

File: spec/predicates/test-results.md
Comment (joshuagl) [inline spec/predicates/test-results.md:85]: Broader question, do we want every predicate to include a summarised copy of the framework's parsing rules? Or just to note where the predicate's parsing rules differ from the framework's parsing rules?

WDYT @in-toto/attestation-maintainers ? We can move this to a separate issue, the discussion should not block this PR.

File: spec/predicates/test-results.md
Comment (joshuagl) [inline spec/predicates/test-results.md:105]: Should `testRun` be required if the `configuration` child field is required?

**Review by joshuagl — COMMENTED** [review]
This looks useful. I had a few questions (inline) about this predicate specifically, and a couple of higher-level questions: one about the relationship between in-toto attestations and the broader in-toto spec, and one about inclusion of attestation framework parsing rules in each predicate.

The higher level questions can be addressed separately and should not block this PR.

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:105]: Yep! I missed this when making configuration required.

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:85]: Speaking with my author of predicates hat on:

It _was_ a bit of work to go back and update each of them as the attestation framework wide rules changed. I could propose a change in ITE-9 that this is only to be used to define deviations from the regular rules.

Comment (joshuagl) [reply spec/predicates/test-results.md:85]: Yeah, maintainability seems much easier with this in one place (that we link to from the predicate?)

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:59]: Updated.

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:31]: I think it's okay to mention them in use cases. I can clarify that it's with in-toto layouts / policy engine. We could similarly add a non in-toto verifier and how it'd do something...

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md]: Updated this one to only link to the spec.

File: spec/predicates/test-results.md
Comment (marcelamelara) [inline spec/predicates/test-results.md:45]: Could you please provide an example for this particular use case where only a subset of tests matter? Is the expectation that this information will be conveyed via the test names in the entries of `passedTests` etc? If so, I think that needs to be expressed more explicitly.

File: spec/predicates/test-results.md
Comment (marcelamelara) [inline spec/predicates/test-results.md:107]: Ditto below.

```suggestion
`testRun.passedTests` _list of strings_, _optional_
```

File: spec/predicates/test-results.md
Comment (marcelamelara) [inline spec/predicates/test-results.md:109]: I would add a comment here (as well as for `warnedTests` and `failedTests`) that the strings in this list are test names, and that their semantics are up to the producer and consumer to decide.

**Review by marcelamelara — CHANGES_REQUESTED** [review]
Thanks @adityasaky !! I really like the simplicity of this predicate. I have several questions/requests for updates to improve clarity.

File: spec/predicates/test-results.md
Comment (TomHennen) [reply spec/predicates/test-results.md:113]: That makes me wonder if instead of having a top level true/false field it would make sense to have top level fields like:

```
testRun.numPassedTests
testRun.numWarnedTests
testRun.numFailedTests
```

Then evaluators are suggested to check:

1. numPassedTests > 0
2. numFailedTests == 0

But can check other stuff if they want too?

Comment (adityasaky) [reply spec/predicates/test-results.md:113]: I think this can be calculated using `len(testRun.{passed,warned,failed}Tests)` correct?

Comment (TomHennen) [reply spec/predicates/test-results.md:113]: Yeah that would work (details would differ depending on language).

If you take this approach would you drop `result`?

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:45]: Added an example with the device types named.

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md:101]: Updated the text here.

File: spec/predicates/test-results.md
Comment (adityasaky) [reply spec/predicates/test-results.md]: I think the result is useful as for simple cases, we can drop the passed, warned, failed lists altogether and just include the result field and the configuration.

Comment (adityasaky) [reply spec/predicates/test-results.md]: Separately, I wonder if we can move all of the testRun fields to the top level? Now that I think about it, I'm not sure we need that separation.

**Comment by adityasaky** [comment]
Does it make more sense for the schema to be as follows? Since `testRun` is now required because of the configuration, we can remove that intermediate field.

```json
{
    "_type": "https://in-toto.io/Statement/v1",
    "subject": [{...}],
    "predicateType": "https://in-toto.io/attestation/test-result/v0.1",
    "predicate": {
        "result": "pass|fail",
        "url": "<URL>",
        "configuration": ["<ResourceDescriptor>", ...],
        "passedTests": ["<TEST_NAME>", ...],
        "warnedTests": ["<TEST_NAME>", ...],
        "failedTests": ["<TEST_NAME>", ...]
    }
}
```

**Comment by adityasaky** [comment]
nit: I'll reorder the fields to move required ones up. Configuration should be first since it's now required.

File: spec/predicates/test-results.md
Comment (TomHennen) [inline spec/predicates/test-results.md:96]: nit: "or `FAILED`"

Comment (adityasaky) [reply spec/predicates/test-results.md]: fixed

**Pull request merged**
